opm/autodiff/BlackoilModelBase_impl.hpp

```diff
@@ -2549,6 +2598,88 @@ namespace detail {
+    {
+        // TODO: for now, we store the voidage rates for all the production wells.
+        // For injection wells, the rates are stored as zero.
+        // Later, more delicate model will be implemented here.
+        // And for the moment, group control can only work for serial running.
+        const int nw = well_state.numWells();
+        const int np = numPhases();
+
+        const Wells* wells = asImpl().wellModel().wellsPointer();
+
+        // we calculate the voidage rate for each well, that means the sum of all the phases.
+        well_voidage_rates.resize(nw, 0);
+        // store the conversion coefficients, while only for the use of injection wells.
+        voidage_conversion_coeffs.resize(nw * np, 1.0);
+
+        int global_number_wells = nw;
+
+#if HAVE_MPI
+        if ( linsolver_.parallelInformation().type() == typeid(ParallelISTLInformation) )
+        {
+            const auto& info =
+                boost::any_cast<const ParallelISTLInformation&>(linsolver_.parallelInformation());
+            global_number_wells = info.communicator().sum(global_number_wells);
+            if ( global_number_wells )
+            {
+                rate_converter_.defineState(reservoir_state, boost::any_cast<const ParallelISTLInformation&>(linsolver_.parallelInformation()));
+            }
+        }
+        else
+#endif
+        {
+            if ( global_number_wells )
+            {
+                rate_converter_.defineState(reservoir_state);
+            }
+        }
+
+        std::vector<double> well_rates(np, 0.0);
+        std::vector<double> convert_coeff(np, 1.0);
+
+
+        if ( !well_voidage_rates.empty() ) {
+            for (int w = 0; w < nw; ++w) {
+                const bool is_producer = wells->type[w] == PRODUCER;
+
+                // not sure necessary to change all the value to be positive
+                if (is_producer) {
+                    std::transform(well_state.wellRates().begin() + np * w,
+                                   well_state.wellRates().begin() + np * (w + 1),
+                                   well_rates.begin(), std::negate<double>());
+
+                    const int fipreg = 0; // Not considering FIP for the moment.
+
+                    rate_converter_.calcCoeff(well_rates, fipreg, convert_coeff);
+                    well_voidage_rates[w] = std::inner_product(well_rates.begin(), well_rates.end(),
+                                                               convert_coeff.begin(), 0.0);
+                } else {
+                    // TODO: Not sure whether will encounter situation with all zero rates
+                    // and whether it will cause problem here.
+                    std::copy(well_state.wellRates().begin() + np * w,
+                              well_state.wellRates().begin() + np * (w + 1),
+                              well_rates.begin());
+                    const int fipreg = 0; // Not considering FIP for the moment.
```

Comment:
How hard would it be to correctly use FIPNUM here and above (assuming that is the correct region number to use)?